### DISCUSSION TITLE
How to convert string to compiled SolidJS in code

### DISCUSSION THREAD

**Discussion by ryanzec** [solidjs/solid-start#1161] [Q&A]
I am looking to create some like : https://formidable.com/open-source/react-live/ : the ability to take a string and in realtime add it in as a component. I am trying to take a look at the https://github.com/solidjs/solid-playground repo but having a hard time figuring out where in that repo is the string being converted into SolidJS code.

Could anyone point me in the right direction for figuring out how to compile a string into SolidJS code and in real time inject it into the current page?

**Comment by atk** [comment]
The playground uses babel and babel-preset-solid in the browser to do that. What do you want to achieve with this?

**Comment by ryanzec** [comment]
> What do you want to achieve with this?

@atk I basically want to be able to create an interactive sandbox that can utilize my own components, similar to what mui does with its docs:
![2023-12-21 02_45_36-React Button component - Material UI](https://github.com/solidjs/solid-start/assets/444206/2f780afb-10d2-4d16-925b-1abed440201a)

(reference: https://mui.com/material-ui/react-button/#basic-button)

In that example the code is editable and update the above content in real time.

**Reply by atk** [reply]
Be sure to use https://babeljs.io/docs/babel-standalone in order to avoid issues when using this in the browser.

**Reply by ryanzec** [reply]
When I try to transform code that has jsx like:
```tsx
const Component = () => {
  return (
    <img alt="avatar image" src={props.src} /> {/* errors when this is include */}
  );
};
```
I get:
```
Uncaught (in promise) TypeError: /index.tsx: _assert is not a function
```
and not sure why.

My call to transform looks like:
```tsx
const tranformed = Babel.transform(code, {
  presets: [
    [babelPresetSolid, { generate: 'dom', hydratable: false }],
    ['typescript', { onlyRemoveTypeImports: true }],
  ],
  filename: 'index.tsx',
});
```

**Reply by atk** [reply]
The JSX is broken. Both children need to be wrapped in `<>...</>`.

If you define _assert, you might get useful error messages pointing this out.

**Reply by ryanzec** [reply]
I only added the comment for the sake of letting you know what piece of code start to cause the error. even a super basic component like this gives the same error
```tsx
const TestComp = () => {
  return (
      <div>test</div>
  );
};
```
I tried defining:
```js
function _assert(...test) {
  console.log(test);
}
```
both in the code being transformed and my code calling the transform and I get the same error.

Looking at were the error is, at the top of that file there is this line in `node_modules/babel-plugin-jsx-dom-expressions/node_modules/@babel/helper-module-imports/lib/import-injector.js`
```js
var _assert = require("assert");
```
So I even tried to manually install the `assert` in my projects root directory and `node_modules/babel-plugin-jsx-dom-expressions` and still got the error.

Outside of example using trivial javascript, I can't find anything for this.

**Reply by atk** [reply]
Try `globalThis._assert = console.assert.bind(console)`.

**Reply by ryanzec** [reply]
same error

**Reply by atk** [reply]
Can you activate breakpoint on error and check in which thread the error occurs?

**Reply by ryanzec** [reply]
It breaks on the main thread
![Screen Shot 2023-12-23 at 1 33 09 PM](https://github.com/solidjs/solid-start/assets/444206/15c10e8c-0476-4c11-9cd8-ca6af026e2be)

**Reply by atk** [reply]
The error seems to happen in dynamic-routes.ts - where does that file come from?

**Reply by ryanzec** [reply]
It is custom code that generates routes dynamically, it is the file the loads the file that has the code that is attempting to do the transformations

**Reply by atk** [reply]
Can you look through the call stack where the call to _assert is made?

**Reply by ryanzec** [reply]
So this is weird, when I deployed this code to vercel, I started to get a different error:
```
SyntaxError: Cannot use import statement outside a module
```
reference link: https://train-ui-sigma.vercel.app/components/playground/default
referenced code: https://github.com/ryanzec/train-ui/blob/main/packages/utils/playground.sandbox.tsx
